### PR TITLE
[FEATURE] Create RepositoryExtension interface

### LINKED ISSUE #329: [FEATURE] Create RepositoryExtension interface
**Is your feature request related to a problem?**

Extension points (see #315) will be (optionally) implemented on Extensions. Eventually these points may need to be communicated to OpenSearch to be registered in the appropriate service or module.

Allowing extensions to inherit from appropriate interfaces depending on their type will match the current Plugin implementation.

**What solution would you like?**

Add a new interface RepositoryExtension with default implementations for extension points matching RepositoryPlugin.java.

### PR DESCRIPTION
**Pull request by kokibas**
### Description
_Describe what this change achieves._

### Issues Resolved

Fixes #329

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/sdk/RepositoryExtension.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ */
+
+public interface RepositoryExtension {
+
+    /**
+     * Returns repository types added by this extension.
+     *
+     * @param env The environment for the local node, which may be used for the local settings and path. repo
+     *
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/RepositoryExtension.java:23]: You need `@param` tags for all of the arguments.

File: src/main/java/org/opensearch/sdk/RepositoryExtension.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * The key of the returned {@link Map} is the type name of the repository and
+     * the value is a factory to construct the {@link Repository} interface.
+     */
+    default Map<String, Repository.Factory> getRepositories(
+            Environment env,
+            NamedXContentRegistry namedXContentRegistry,
+            ClusterService clusterService,
+            RecoverySettings recoverySettings
+    ) {
+        return Collections.emptyMap();
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/RepositoryExtension.java:33]: Please add tests to `TestExtensionInterfaces` for these return values.

**Review by dbwiddis — CHANGES_REQUESTED** [review]

**Comment by kokibas** [comment]
I apologize, but I am very confused, I would like to make a new one and try again.If that s okay.

**Comment by dbwiddis** [comment]
Sometimes that happens when you're learning git.  My very first github PR I ended up deleting and trying again. :)

**Pull request closed without merging**